File: trip-kit-booking/src/main/java/com/skedgo/tripkit/booking/viewmodel/AuthenticationViewModelImpl.java
package com.skedgo.tripkit.booking.viewmodel;

import com.skedgo.tripkit.common.rx.Var;
import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;

public final class AuthenticationViewModelImpl implements AuthenticationViewModel {
  private final Var<Boolean> isSuccessful = Var.create(false);
  private final Var<String> url = Var.create();

  @Override
  public Var<String> url() {
    return url;
  }

  @Override
  public Observable<Boolean> isSuccessful() {
    return isSuccessful.observe().toObservable();
  }

  @Override
  public Observable<Boolean> verify(final Object param) {
    return Observable
        .create(new ObservableOnSubscribe<Boolean>() {
          @Override
          public void subscribe(ObservableEmitter<Boolean> subscriber) {
            final String url = String.valueOf(param);
            subscriber.onNext(url != null && url.contains("skedgo.com"));
            subscriber.onComplete();
          }
        })
        .doOnNext(isSuccessful);
  }
}

File: trip-kit-booking/src/main/java/com/skedgo/tripkit/booking/viewmodel/AuthenticationViewModelImpl.kt
package com.skedgo.tripkit.booking.viewmodel

import io.reactivex.Observable
import io.reactivex.subjects.BehaviorSubject

class AuthenticationViewModelImpl : AuthenticationViewModel {
    private val isSuccessful: BehaviorSubject<Boolean> = BehaviorSubject.createDefault(false)
    private val url: BehaviorSubject<String> = BehaviorSubject.create()

    override fun url(): BehaviorSubject<String> {
        return url
    }

    override fun isSuccessful(): Observable<Boolean> {
        // Directly return the BehaviorSubject as an Observable
        return isSuccessful.hide()
    }

    override fun verify(param: Any): Observable<Boolean> {
        return Observable.create { emitter ->
            val urlString = param.toString()
            val isUrlValid = urlString.contains("skedgo.com")
            emitter.onNext(isUrlValid)
            emitter.onComplete()
        }.doOnNext { isSuccessful.onNext(it) }
    }
}